File: sdks/ts/memory/src/ingest/state-store.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Pipeline state store interface and file-based implementation.
 *
 * Tracks the progress of each document through the ingest pipeline so that
 * crash recovery can resume from the last completed stage rather than
 * re-processing from scratch or (worse) permanently skipping the document.
 *
 * The FilePipelineStateStore persists state as JSON files inside the brain's
 * Store at `raw/.pipeline-state/{hash}.json`.
 */

import type { Store } from '../store/index.js'
import { isNotFound, toPath } from '../store/index.js'

/** Ordered stages a document passes through during ingestion. */
export type PipelineStage =
  | 'received'
  | 'stored'
  | 'chunked'
  | 'embedded'
  | 'indexed'
  | 'completed'
  | 'failed'

/** Terminal stages that indicate a document is no longer in-flight. */
const TERMINAL_STAGES: ReadonlySet<PipelineStage> = new Set(['completed', 'failed'])

/** A single document's pipeline state record. */
export type PipelineStateEntry = {
  readonly documentHash: string
  readonly brainId: string
  readonly stage: PipelineStage
  readonly retryCount: number
  readonly lastError?: string
  readonly createdAt: Date
  readonly updatedAt: Date
  readonly completedAt?: Date
}

/**
 * Persistence interface for pipeline state tracking. Implementations must
 * be safe for sequential use within a single pipeline run.
 */
export type PipelineStateStore = {
  /** Retrieve state for a document. Returns undefined when no record exists. */
  get(documentHash: string, signal?: AbortSignal): Promise<PipelineStateEntry | undefined>

  /** Create or overwrite state for a document. */
  set(entry: PipelineStateEntry, signal?: AbortSignal): Promise<void>

  /** Return all entries not in a terminal stage, optionally filtered by brainId. */
  listIncomplete(brainId: string, signal?: AbortSignal): Promise<readonly PipelineStateEntry[]>

  /** Remove the state record for a document. No-op if the record does not exist. */
  delete(documentHash: string, signal?: AbortSignal): Promise<void>
}

const DEFAULT_STATE_PREFIX = 'raw/.pipeline-state'

/**
 * Validates that a document hash contains only lowercase hex characters and is
 * exactly 64 characters long (SHA-256 hex). Rejects path traversal attempts.
 */
const DOCUMENT_HASH_PATTERN = /^[a-f0-9]{64}$/

const assertDocumentHash = (documentHash: string): void => {
  if (!DOCUMENT_HASH_PATTERN.test(documentHash)) {
    throw new Error(`Invalid document hash: "${documentHash}"`)
  }
}

const statePath = (prefix: string, documentHash: string) => toPath(`${prefix}/${documentHash}.json`)

/** JSON wire format for serialisation. Dates are stored as ISO strings. */
type SerializedEntry = {
  readonly documentHash: string
  readonly brainId: string
  readonly stage: PipelineStage
  readonly retryCount: number
  readonly lastError?: string
  readonly createdAt: string
  readonly updatedAt: string
  readonly completedAt?: string
}

const serialize = (entry: PipelineStateEntry): Buffer => {
  const wire: SerializedEntry = {
    documentHash: entry.documentHash,
    brainId: entry.brainId,
    stage: entry.stage,
    retryCount: entry.retryCount,
    ...(entry.lastError !== undefined ? { lastError: entry.lastError } : {}),
    createdAt: entry.createdAt.toISOString(),
    updatedAt: entry.updatedAt.toISOString(),
    ...(entry.completedAt !== undefined ? { completedAt: entry.completedAt.toISOString() } : {}),
  }
  return Buffer.from(JSON.stringify(wire, null, 2), 'utf8')
}

const deserialize = (buf: Buffer): PipelineStateEntry => {
  const raw: unknown = JSON.parse(buf.toString('utf8'))
  const wire = raw as SerializedEntry
  return {
    documentHash: wire.documentHash,
    brainId: wire.brainId,
    stage: wire.stage,
    retryCount: wire.retryCount,
    ...(wire.lastError !== undefined ? { lastError: wire.lastError } : {}),
    createdAt: new Date(wire.createdAt),
    updatedAt: new Date(wire.updatedAt),
    ...(wire.completedAt !== undefined ? { completedAt: new Date(wire.completedAt) } : {}),
  }
}

/** Options for constructing a FilePipelineStateStore. */
export type FilePipelineStateStoreOptions = {
  /** Brain Store backend used for persistence. */
  readonly store: Store
  /**
   * Path prefix inside the store where state files are written.
   * Defaults to `raw/.pipeline-state`.
   */
  readonly prefix?: string
}

/**
 * File-based pipeline state store. Persists each document's state as a JSON
 * file at `{prefix}/{documentHash}.json` using the brain Store interface.
 * The default prefix is `raw/.pipeline-state`.
 */
export class FilePipelineStateStore implements PipelineStateStore {
  private readonly store: Store
  private readonly prefix: string

  constructor(store: Store)
  constructor(opts: FilePipelineStateStoreOptions)
  constructor(storeOrOpts: Store | FilePipelineStateStoreOptions) {
    if ('read' in storeOrOpts) {
      this.store = storeOrOpts
      this.prefix = DEFAULT_STATE_PREFIX
    } else {
      this.store = storeOrOpts.store
      this.prefix = storeOrOpts.prefix ?? DEFAULT_STATE_PREFIX
    }
  }

  async get(documentHash: string, _signal?: AbortSignal): Promise<PipelineStateEntry | undefined> {
    assertDocumentHash(documentHash)
    try {
      const buf = await this.store.read(statePath(this.prefix, documentHash))
      return deserialize(buf)
    } catch (err: unknown) {
      if (isNotFound(err)) return undefined
      throw err
    }
  }

  async set(entry: PipelineStateEntry, _signal?: AbortSignal): Promise<void> {
    assertDocumentHash(entry.documentHash)
    await this.store.write(statePath(this.prefix, entry.documentHash), serialize(entry))
  }

  async listIncomplete(brainId: string, _signal?: AbortSignal): Promise<readonly PipelineStateEntry[]> {
    const dir = toPath(this.prefix)
    const exists = await this.store.exists(dir)
    if (!exists) return []

    const files = await this.store.list(dir, { recursive: false, glob: '*.json' })
    const entries: PipelineStateEntry[] = []

    for (const file of files) {
      if (file.isDir) continue
      const buf = await this.store.read(file.path)
      const entry = deserialize(buf)
      if (entry.brainId === brainId && !TERMINAL_STAGES.has(entry.stage)) {
        entries.push(entry)
      }
    }

    return entries
  }

  async delete(documentHash: string, _signal?: AbortSignal): Promise<void> {
    assertDocumentHash(documentHash)
    try {
      await this.store.delete(statePath(this.prefix, documentHash))
    } catch (err: unknown) {
      if (isNotFound(err)) return
      throw err
    }
  }
}

File: go/ingest/state_store.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"time"

	"github.com/jeffs-brain/memory/go/brain"
)

// PipelineStage represents the ordered stages a document passes through
// during ingestion. Terminal stages are "completed" and "failed".
type PipelineStage string

const (
	StageReceived  PipelineStage = "received"
	StageStored    PipelineStage = "stored"
	StageChunked   PipelineStage = "chunked"
	StageEmbedded  PipelineStage = "embedded"
	StageIndexed   PipelineStage = "indexed"
	StageCompleted PipelineStage = "completed"
	StageFailed    PipelineStage = "failed"
)

// IsTerminal reports whether the stage represents a final state that
// should not appear in ListIncomplete results.
func (s PipelineStage) IsTerminal() bool {
	return s == StageCompleted || s == StageFailed
}

// PipelineStateEntry tracks a single document's progress through the
// ingest pipeline.
type PipelineStateEntry struct {
	DocumentHash string        `json:"documentHash"`
	BrainID      string        `json:"brainId"`
	Stage        PipelineStage `json:"stage"`
	RetryCount   int           `json:"retryCount"`
	LastError    string        `json:"lastError,omitempty"`
	CreatedAt    time.Time     `json:"createdAt"`
	UpdatedAt    time.Time     `json:"updatedAt"`
	CompletedAt  *time.Time    `json:"completedAt,omitempty"`
}

// PipelineStateStore persists pipeline state for crash recovery. Both
// file-based and PostgreSQL implementations satisfy this interface.
type PipelineStateStore interface {
	// Get retrieves the state entry for the given document hash. Returns
	// nil without error when no record exists.
	Get(ctx context.Context, documentHash string) (*PipelineStateEntry, error)

	// Set creates or overwrites the state entry for a document.
	Set(ctx context.Context, entry PipelineStateEntry) error

	// ListIncomplete returns all entries for the given brainID whose stage
	// is not terminal (not "completed" or "failed").
	ListIncomplete(ctx context.Context, brainID string) ([]PipelineStateEntry, error)

	// Delete removes the state entry for the given document hash. Does not
	// return an error if the entry does not exist.
	Delete(ctx context.Context, documentHash string) error
}

const defaultStatePrefix = "raw/.pipeline-state"

// documentHashPattern validates that a document hash is a 64-character
// lowercase hex string (SHA-256). Rejects path traversal attempts.
var documentHashPattern = regexp.MustCompile(`^[a-f0-9]{64}$`)

func validateDocumentHash(documentHash string) error {
	if !documentHashPattern.MatchString(documentHash) {
		return fmt.Errorf("ingest: invalid document hash %q", documentHash)
	}
	return nil
}

func statePath(prefix, documentHash string) brain.Path {
	return brain.Path(fmt.Sprintf("%s/%s.json", prefix, documentHash))
}

// FilePipelineStateStoreConfig configures the file-based state store.
type FilePipelineStateStoreConfig struct {
	// Store is the brain Store backend used for persistence.
	Store brain.Store
	// Prefix is the path prefix inside the store where state files are written.
	// Defaults to "raw/.pipeline-state" when empty.
	Prefix string
}

// FilePipelineStateStore implements PipelineStateStore by persisting each
// document's state as a JSON file at {prefix}/{hash}.json using the
// brain.Store interface. The default prefix is "raw/.pipeline-state".
type FilePipelineStateStore struct {
	store  brain.Store
	prefix string
}

// NewFilePipelineStateStore creates a file-based state store backed by the
// given brain.Store using the default prefix.
func NewFilePipelineStateStore(store brain.Store) *FilePipelineStateStore {
	return &FilePipelineStateStore{store: store, prefix: defaultStatePrefix}
}

// NewFilePipelineStateStoreWithConfig creates a file-based state store with
// explicit configuration, including a custom path prefix.
func NewFilePipelineStateStoreWithConfig(cfg FilePipelineStateStoreConfig) *FilePipelineStateStore {
	prefix := cfg.Prefix
	if prefix == "" {
		prefix = defaultStatePrefix
	}
	return &FilePipelineStateStore{store: cfg.Store, prefix: prefix}
}

func (s *FilePipelineStateStore) Get(ctx context.Context, documentHash string) (*PipelineStateEntry, error) {
	if err := validateDocumentHash(documentHash); err != nil {
		return nil, err
	}
	data, err := s.store.Read(ctx, statePath(s.prefix, documentHash))
	if err != nil {
		if errors.Is(err, brain.ErrNotFound) {
			return nil, nil
		}
		return nil, fmt.Errorf("ingest: state get %s: %w", documentHash, err)
	}
	var entry PipelineStateEntry
	if err := json.Unmarshal(data, &entry); err != nil {
		return nil, fmt.Errorf("ingest: state decode %s: %w", documentHash, err)
	}
	return &entry, nil
}

func (s *FilePipelineStateStore) Set(ctx context.Context, entry PipelineStateEntry) error {
	if err := validateDocumentHash(entry.DocumentHash); err != nil {
		return err
	}
	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return fmt.Errorf("ingest: state encode %s: %w", entry.DocumentHash, err)
	}
	if err := s.store.Write(ctx, statePath(s.prefix, entry.DocumentHash), data); err != nil {
		return fmt.Errorf("ingest: state set %s: %w", entry.DocumentHash, err)
	}
	return nil
}

func (s *FilePipelineStateStore) ListIncomplete(ctx context.Context, brainID string) ([]PipelineStateEntry, error) {
	dir := brain.Path(s.prefix)
	exists, err := s.store.Exists(ctx, dir)
	if err != nil {
		return nil, fmt.Errorf("ingest: state list check dir: %w", err)
	}
	if !exists {
		return nil, nil
	}

	files, err := s.store.List(ctx, dir, brain.ListOpts{Recursive: false, Glob: "*.json"})
	if err != nil {
		return nil, fmt.Errorf("ingest: state list: %w", err)
	}

	entries := make([]PipelineStateEntry, 0, len(files))
	for _, file := range files {
		if file.IsDir {
			continue
		}
		data, readErr := s.store.Read(ctx, file.Path)
		if readErr != nil {
			if errors.Is(readErr, brain.ErrNotFound) {
				continue
			}
			return nil, fmt.Errorf("ingest: state list read %s: %w", file.Path, readErr)
		}
		var entry PipelineStateEntry
		if decodeErr := json.Unmarshal(data, &entry); decodeErr != nil {
			return nil, fmt.Errorf("ingest: state list decode %s: %w", file.Path, decodeErr)
		}
		if entry.BrainID == brainID && !entry.Stage.IsTerminal() {
			entries = append(entries, entry)
		}
	}

	return entries, nil
}

func (s *FilePipelineStateStore) Delete(ctx context.Context, documentHash string) error {
	if err := validateDocumentHash(documentHash); err != nil {
		return err
	}
	err := s.store.Delete(ctx, statePath(s.prefix, documentHash))
	if err != nil {
		if errors.Is(err, brain.ErrNotFound) {
			return nil
		}
		return fmt.Errorf("ingest: state delete %s: %w", documentHash, err)
	}
	return nil
}

var _ PipelineStateStore = (*FilePipelineStateStore)(nil)
